snowflake.go

```diff
@@ -101,40 +125,55 @@ func NewNode(node int64) (*Node, error) {
 	nodeShift = StepBits
 
 	return &Node{
-		time: 0,
+		last: 0,
 		node: node,
 		step: 0,
 	}, nil
 }
 
 // Generate creates and returns a unique snowflake ID
+// use GenerateID to generate snowflakes with error checking
 func (n *Node) Generate() ID {
+	id, _ := n.doGenerateID()
+	return id
+}
+
+// GenerateID attempts to create and return a unique snowflake ID
+// If backwards clock drift is observed, a BackwardsTimeError will be returned
+func (n *Node) GenerateID() (ID, error) {
+	return n.doGenerateID()
+}
 
+func (n *Node) doGenerateID() (r ID, err error) {
 	n.mu.Lock()
 
 	now := time.Now().UnixNano() / 1000000
 
-	if n.time == now {
+	if n.last == now {
 		n.step = (n.step + 1) & stepMask
 
 		if n.step == 0 {
-			for now <= n.time {
+			for now <= n.last {
 				now = time.Now().UnixNano() / 1000000
 			}
 		}
 	} else {
 		n.step = 0
 	}
 
-	n.time = now
-
-	r := ID((now-Epoch)<<timeShift |
+	r = ID((now-Epoch)<<timeShift |
 		(n.node << nodeShift) |
 		(n.step),
 	)
 
+	if now < n.last {
+		err = NewBackwardsTimeError(n.last - now)
+	}
+
+	n.last = now
```

Comment:
The only thing I'm not sure about now is this. Technically if you care about clock drift, you will move `last` backwards here. Which means even if you don't wait `BackwardsTimeError.Offset()` the next ID generated will look valid.

This might deserve a bit more splitting of the code to support both scenarios.

Reply:
I think _maybe_ we can ignore this... it seems contradictory to care about clock drift, but not handle the first error appropriately. I'll wait to get some feedback.